libursa/src/bn/rust.rs

```diff
@@ -0,0 +1,632 @@
+use serde::de::{Deserialize, Deserializer, Visitor, Error as DError};
+
+use std::error::Error;
+use std::fmt;
+use std::cmp::Ord;
+use std::cmp::Ordering;
+
+pub struct BigNumberContext;
+
+pub struct BigNumber {
+    bn: BigInt
+}
+
+impl BigNumber {
+    pub fn new_context() -> Result<BigNumberContext, UrsaCryptoError> {
+        Ok(BigNumberContext{})
+    }
+
+    pub fn new() -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber {
+            bn: BigInt::zero()
+        })
+    }
+
+    pub fn generate_prime(size: usize) -> Result<BigNumber, UrsaCryptoError> {
+        match prime::new(size)?.to_bigint() {
+            Some(bn) => Ok(BigNumber { bn }),
+            None => Err(UrsaCryptoError::InvalidStructure("Unable to generate prime".to_string()))
+        }
+    }
+
+    pub fn generate_safe_prime(size: usize) -> Result<BigNumber, UrsaCryptoError> {
+        match safe_prime::new(size)?.to_bigint() {
+            Some(bn) => Ok(BigNumber { bn }),
+            None => Err(UrsaCryptoError::InvalidStructure("Unable to generate safe prime".to_string()))
+        }
+    }
+
+    pub fn generate_prime_in_range(start: &BigNumber, end: &BigNumber) -> Result<BigNumber, UrsaCryptoError> {
+        let mut res;
+        let mut iteration = 0;
+        let mut rng = OsRng::new()?;
+        let start = match start.bn.to_biguint() {
+            Some(bn) => bn,
+            None => return Err(UrsaCryptoError::InvalidStructure(format!("Invalid number for 'start': {:?}", start)))
+        };
+        let end = match end.bn.to_biguint() {
+            Some(bn) => bn,
+            None => return Err(UrsaCryptoError::InvalidStructure(format!("Invalid number for 'end': {:?}", end)))
+        };
+
+        let bits = (&end -  &start).bits();
+        let mask = (BigUint::from(3u8) << (bits - 2)) | BigUint::one();
+
+        loop {
+            res = rng.gen_biguint_range(&start, &end);
+            res |= &mask;
+
+            if prime::check(&res) {
+                debug!("Found prime in {} iteration", iteration);
+                break;
+            }
+            iteration += 1;
+        }
+
+        match res.to_bigint() {
+            Some(bn) => Ok(BigNumber{bn}),
+            None => Err(UrsaCryptoError::InvalidStructure("Unable to generate prime in range".to_string()))
+        }
+    }
+
+    pub fn is_prime(&self, _ctx: Option<&mut BigNumberContext>) -> Result<bool, UrsaCryptoError> {
+        if self.is_negative() {
+            Ok(false)
+        } else {
+            match self.bn.to_biguint() {
+                Some(bn) => Ok(prime::check(&bn)),
+                None => Err(UrsaCryptoError::InvalidStructure("An error in is_prime".to_string()))
+            }
+        }
+    }
+
+    pub fn is_safe_prime(&self, _ctx: Option<&mut BigNumberContext>) -> Result<bool, UrsaCryptoError> {
+        if self.bn < BigInt::zero() {
+            Ok(false)
+        } else {
+            match self.bn.to_biguint() {
```

Comment:
Can't this logic be abstracted in a macro taking `prime` or `safe_prime` and an error message as an argument?